appengine/genetics/js/visualization.js

```diff
@@ -226,319 +303,525 @@ Genetics.Visualization.resetChartData_ = function() {
+    if (playerStats['pickFightDiv']) {
+      playerStats['pickFightDiv'].style.width = pickFightPercent + '%';
+    }
+    var proposeMatePercent = (100 *
+        Genetics.Visualization.proposeMateOwners_[playerId] / mouseCount) || 0;
+    if (playerStats['proposeMateDiv']) {
+      playerStats['proposeMateDiv'].style.width = proposeMatePercent + '%';
+    }
+    var acceptMatePercent = (100 *
+        Genetics.Visualization.acceptMateOwners_[playerId] / mouseCount) || 0;
+    if (playerStats['acceptMateDiv']) {
+      playerStats['acceptMateDiv'].style.width = acceptMatePercent + '%';
+    }
+    playerStats['td'].title = 'pickFight ' +
+        Math.round(pickFightPercent * 100) / 100 + '%\nproposeMate ' +
+        Math.round(proposeMatePercent * 100) / 100 + '%\nacceptMate ' +
+        Math.round(acceptMatePercent * 100) / 100 + '%';
+  }
+};
 
 /**
- * Visualize the current state of the cage simulation (Genetics.Cage).
+ * Redraw game charts.
+ * @param {boolean=} force Whether to draw charts even if they are not visible.
  * @private
  */
-Genetics.Visualization.display_ = function() {
-  // TODO(kozbial) draw current state of the cage.
-  if (Genetics.Visualization.chartsUpdated_) {
+Genetics.Visualization.drawCharts_ = function(opt_force) {
+  if (Genetics.Visualization.chartsNeedUpdate_ &&
+      (opt_force || Genetics.Visualization.areChartsVisible_)) {
+    // Redraw the charts in the charts tab.
     Genetics.Visualization.populationChartWrapper_.draw();
     Genetics.Visualization.pickFightChartWrapper_.draw();
     Genetics.Visualization.proposeMateChartWrapper_.draw();
     Genetics.Visualization.acceptMateChartWrapper_.draw();
-    Genetics.Visualization.chartsUpdated_ = false;
+
+    Genetics.Visualization.chartsNeedUpdate_ = false;
   }
 };
 
 /**
- * Count for events to keep track of time passed for charts.
- * @type {number}
+ * The current round number.
+ * @private {number}
+ */
+Genetics.Visualization.roundNumber_ = 0;
+
+/**
+ * The index of the next event to be processed or 0 if events are not
+ * removed from EVENTS queue.
+ * @private {number}
  */
-Genetics.Visualization.eventNumber = 0;
+Genetics.Visualization.eventIndex_ = 0;
 
 /**
  * Process events in Cage events queue.
  * @private
  */
 Genetics.Visualization.processCageEvents_ = function() {
   // Handle any queued events.
-  var getMouseName = Genetics.Visualization.getMouseName;
-  while (Genetics.Cage.EVENTS.length) {
-    var event = Genetics.Cage.EVENTS.shift();
+  var getMouseName = Genetics.Visualization.getMouseName_;
+  while (Genetics.Visualization.eventIndex_ < Genetics.Cage.Events.length) {
+    var event = Genetics.Cage.historyPreserved ?
+        Genetics.Cage.Events[Genetics.Visualization.eventIndex_++] :
+        Genetics.Cage.Events.shift();
+
+    var mouse = (event['ID'] !== null) ?
+        Genetics.Visualization.mice_[event['ID']] : null;
+    var opponent = (event['OPT_OPPONENT'] != null) ?
```

Comment:
Actually, I realize that I was meaning to see if was not undefined. I've changed to format
`var mouse = event['ID'] && Genetics.Visualization.mice_[event['ID']];`
Since allowing them to be undefined works just as well for me.